plugins/dell-kestrel/fu-dell-kestrel-ec.c

```diff
@@ -0,0 +1,1052 @@
+/*
+ * Copyright 2024 Dell Technologies
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR MIT
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-dell-kestrel-common.h"
+
+typedef struct __attribute__((packed)) { /* nocheck:blocked */
+	struct FuDellKestrelV2DockInfoHeader {
+		guint8 total_devices;
+		guint8 first_index;
+		guint8 last_index;
+	} header;
+	struct FuDellKestrelEcQueryEntry {
+		struct FuDellKestrelV2EcAddrMap {
+			guint8 location;
+			guint8 device_type;
+			guint8 sub_type;
+			guint8 arg;
+			guint8 instance;
+		} ec_addr_map;
+		union {
+			guint32 version_32;
+			guint8 version_8[4];
+		} __attribute__((packed)) version; /* nocheck:blocked */
+	} devices[20];
+} FuDellKestrelDockInfoStructure;
+
+/* Private structure */
+struct _FuDellKestrelEc {
+	FuDevice parent_instance;
+	FuStructDellKestrelDockData *dock_data;
+	FuDellKestrelDockInfoStructure *dock_info;
+	FuDellDockBaseType base_type;
+	FuDellKestrelDockSku base_sku;
+	guint64 blob_version_offset;
+	gboolean dock_lock_state;
+};
+
+G_DEFINE_TYPE(FuDellKestrelEc, fu_dell_kestrel_ec, FU_TYPE_HID_DEVICE)
+
+static struct FuDellKestrelEcQueryEntry *
+fu_dell_kestrel_ec_dev_entry(FuDevice *device,
+			     FuDellKestrelEcDevType dev_type,
+			     guint8 sub_type,
+			     guint8 instance)
+{
+	FuDellKestrelEc *self = FU_DELL_KESTREL_EC(device);
+
+	for (guint i = 0; i < self->dock_info->header.total_devices; i++) {
+		if (self->dock_info->devices[i].ec_addr_map.device_type != dev_type)
+			continue;
+		if (sub_type != 0 && self->dock_info->devices[i].ec_addr_map.sub_type != sub_type)
+			continue;
+
+		/* vary by instance index */
+		if (dev_type == FU_DELL_KESTREL_EC_DEV_TYPE_PD &&
+		    self->dock_info->devices[i].ec_addr_map.instance != instance)
+			continue;
+
+		return &self->dock_info->devices[i];
+	}
+	return NULL;
+}
+
+gboolean
+fu_dell_kestrel_ec_is_dev_present(FuDevice *device,
+				  FuDellKestrelEcDevType dev_type,
+				  guint8 sub_type,
+				  guint8 instance)
+{
+	return fu_dell_kestrel_ec_dev_entry(device, dev_type, sub_type, instance) != NULL;
+}
+
+const gchar *
+fu_dell_kestrel_ec_devicetype_to_str(FuDellKestrelEcDevType dev_type,
+				     guint8 sub_type,
+				     guint8 instance)
+{
```

Comment:
Can this whole thing be a rs with a `ToString`?  I woudl think you can make it an rs that has a 16 bit enum or something like that.

Reply:
The name will become `xxx-yyy` in lowercase, and this func still exists as a mapping table to retrieve the `str`. More code but little gains.